### PR TITLE
BitLocker OpenCL new attack modes

### PR DESCRIPTION
**Pull request by e-ago**
bitlocker2john:
- Now it can extract also the Recovery Key from the input image
- It produces 4 output hash files: 
    - hash_user_pass.txt: hash with version number $0$, needed to start the standard (fast) User Password attack
    - hash_user_pass_mac.txt: hash with version number $1$, needed to start the User Password attack with MAC verification (slow performance)
    - hash_recv_pass.txt: hash with version number $2$, needed to start the standard (fast) Recovery Password attack
    - hash_recv_pass_mac.txt: hash with version number $3$, needed to start the Recovery Password attack with MAC verification (slow performance)
- The Recovery Password extraction has been tested only with images of memory devices encrypted using the User Password authentication method. It would be interesting to run a test on a memory device encrypted with, for example, a Smart Card
- The MAC is always stored in the hash

opencl_bitlocker_fmt_plug.c and bitlocker_kernel.cl:
- Code reworked, variables renamed
- New attack modes: User Password with MAC verification, Recovery Password and Recovery Password with MAC verification
- OpenCL kernel fed with passwords as integers (instead of chars)
- Maximum password length increased to 55 chars (in order to accept Recovery Passwords)
- Autotune time took forever with values (1000000000 : 10000000000ULL) and returned this error on GPU NVIDIA Tesla K80:
“
gws:   8388608	 16014Kc/s16792301862912 rounds/s 523.816ms per crypt_all()+
Hardware resources exhausted
”
For this reason I reduced values to (10000 : 100000000ULL) to avoid autotune hangs on big GWS and LWS

bitlocker_common.c:
- New constants added to share with the BitLocker CPU format the same version numbers
- New buffer in “bitlocker_custom_salt” to store the MAC value when parsing the input hash

bitlocker_common_plug.c:
- New test cases added (OpenCL format only) for different attack types: User Password + MAC, Recovery Password and Recovery Password + MAC
- bitlocker_common_valid() and bitlocker_common_get_salt() functions updated according to the new attack modes

Additional comments:
- Tested on a GPU NVIDIA Tesla K80 and CPU i7 quad-core
- Extensive performance tests not done yet
- Recovery Password dictionary and hashes taken from https://github.com/e-ago/bitcracker 
- Currently, produce a dictionary with all the possible Recovery Passwords would required hundreds of Terabytes
- Recovery Password performance are similar to the User Password performance
- The format of a Recovery Password is XXXXXX-XXXXXX-XXXXXX-XXXXXX-XXXXXX-XXXXXX-XXXXXX-XXXXXX where 0 <= X <= 9 and each XXXXXX must be divisible by 11. Thus, the mask format in case of Recovery Password would be:

-mask=[0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9][-][0-9][0-9][0-9][0-9][0-9][0-9]

Or

-mask=?d?d?d?d?d?d[-]?d?d?d?d?d?d[-]?d?d?d?d?d?d[-]?d?d?d?d?d?d[-]?d?d?d?d?d?d[-]?d?d?d?d?d?d[-]?d?d?d?d?d?d[-]?d?d?d?d?d?d

Is it possible to check if every 6-digit number is divisible by 11?

### CODE REVIEW

**Comment by e-ago** [comment]
I've merged my repo with the original repo and it says to me that only *bitlocker* files have been changed. If you see something weird with this pull request let me know and I'll try to fix it or I will reopen the pull request

**Comment by solardiz** [comment]
> Is it possible to check if every 6-digit number is divisible by 11?

With external mode, yes. It could provide a filter() for that and be used on top of another mode such as mask, or it could generate suitable candidate passwords on its own. However, your example mask would produce way too many combinations to ever check, even if we skip 10 of 11 combinations for every 6 digit portion, so I guess I am missing something? Perhaps some of those XXXXXX'es would somehow be known?

**Comment by e-ago** [comment]
@solardiz You're right, the problem is that the possible recovery passwords are not infinite but they are definitely a lot! The range for every XXXXXX number is between 0 <= and <= 720885 and every XXXXXX must be divisible by 11. I'm searching for a way to build a "smart" dictionary or retrieve some additional info to reduce the amount of possible combinations 

@kholia CPU BitLocker format didn't pass the test because I changed the hash parsing function: the 60 chars of VMK now have been split into 16 bytes for MAC and 44 bytes for VMK.

**Comment by kholia** [comment]
@e-ago Great work, and thanks for this PR :+1: 

The CPU BitLocker is failing the CI testing because new unsupported hashes were added to `bitlocker_tests`. Trying to restrict these new hashes to the OpenCL format using the `#ifdef HAVE_OPENCL` hack won't work at all.

@e-ago I can add support for cracking the new hash types to the CPU format. To do so, I will require a high level description of the steps involved (something like -> derive key from password using the KDF, do AES CCM decryption, use algorithm (XYZ) to derive a MAC from (ABC) data, compare this derived MAC to the MAC present in the input hash at (POS) position). I have requested this information before but without any luck ;(

Some quick comments,

1. There are trailing spaces present in `bitlocker_fmt_plug.c`. Please fix.

2. Instead of saying "int xyz=1", say "int xyz = 1" (added spaces around operators).

I have just started looking at this PR.

**Comment by solardiz** [comment]
> to reduce the amount of possible combinations

Another thing to consider: is the distribution uniform? Maybe the keyspace will stay too large, but a non-negligible percentage of the keys will fall in a portion of the keyspace that is feasible to search.

For example, how are the numbers limited to the range ending 720885? If it's e.g. modulo division of some potentially only slightly larger random number by 720886 and the original full range of that random number isn't a multiple of 720886, that creates bias. This alone is probably not enough for our needs, but I merely provide this as an example.

**Comment by solardiz** [comment]
Oh, I see that 720885 is (2^16-1)*11. So I guess it's arrived at by multiplying a non-biased(?) number in the range of 0 to 2^16-1 by 11. How are those 16-bit numbers obtained, though? Are they possibly obtained from a non-cryptographic PRNG with a predictable and/or small seed or internal state?

**Comment by kholia** [comment]
@e-ago I am going to first review, test, and commit `bitlocker2john.c`. It would be nice if you can squash your commits and remove the changes related to `bitlocker2john.c` file from this PR.

**Comment by kholia** [comment]
Before this change,

```
$ ../run/bitlocker2john -i ~/libbde/samples/BitLockerToGo.raw -o x
Opening file ~/libbde/samples/BitLockerToGo.raw

Signature found at 0x0008ef7c
Version: 0 
Invalid version, looking for a signature with valid version...

Signature found at 0x02100000
Version: 2 (Windows 7 or later)
VMK entry found at 0x021000c2
VMK encrypted with user password found!
~/libbde/samples/BitLockerToGo.raw result hash:
$bitlocker$0$16$9079aaee7be0923b529f069012f30b13$1048576$12$40ea50c2b79fd20103000000$60$00000000000000000000000000000000209d495ed0b61eda9c83a79f0abb3f598853b6f89cdffd3b5b66b90b321b822c90c8ef5dac464ef6edd06881
```

After this change,

```
$ ../run/bitlocker2john -i ~/libbde/samples/BitLockerToGo.raw 

---------> bitlocker2john hash extractor <---------
Opening file ~/libbde/samples/BitLockerToGo.raw

Signature found at 0x0008ef7c
Version: 0 
Invalid version, looking for a signature with valid version...

Signature found at 0x02100000
Version: 2 (Windows 7 or later)

VMK entry found at 0x021000c2
VMK encrypted with user password found!
VMK encrypted with AES-CCM

VMK entry found at 0x021001a2
VMK encrypted with Recovery key found!
VMK encrypted with AES-CCM
Error creating ./(null) : Bad address

Error while parsing input device image
```

I can probably fix this part. My cleaned up code is at https://github.com/kholia/JohnTheRipper/tree/BitLocker-NG location.

I believe that hashes should only be printed to stdout, and not written to any output files. Removing this unneeded functionality will improve the robustness of this code. I will attempt to do so.

**Comment by kholia** [comment]
> I can probably fix this part.

Fixed in https://github.com/kholia/JohnTheRipper/tree/BitLocker-NG in a separate commit. The generated hashes look fine.

I will try to cleanup the code a bit more.

**Comment by kholia** [comment]
> To do so, I will require a high level description of the steps involved.

Specifically, what happens with `d_macIV`, `d_cMacIV`, and `d_mac` values in the OpenCL kernel? I see that some of them are undergoing AES encryption in the kernel but the exact high-level understanding eludes me.

**Comment by kholia** [comment]
@e-ago https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History should be helpful in learning Git. It shows how to squash, and amend commits. Combine these two things with `git pull --rebase <upstream remote name> bleeding-jumbo` and `git reset --hard HEAD~<N<` commands carefully, and you should be all set.

**Comment by kholia** [comment]
PR https://github.com/magnumripper/JohnTheRipper/pull/3023 is about merging the `bitlocker2john.c` stuff first. Once that is done, I will focus on getting the `bitlocker_common_plug.c` changes merged.

**Comment by kholia** [comment]
> Trying to restrict these new hashes to the OpenCL format using the `#ifdef HAVE_OPENCL` hack won't work at all.

This hack works when combined with the following changes,

``` diff
diff --git a/src/bitlocker_common_plug.c b/src/bitlocker_common_plug.c
index b51d7896f..4e36f3a27 100644
--- a/src/bitlocker_common_plug.c
+++ b/src/bitlocker_common_plug.c
@@ -121,9 +136,18 @@ void *bitlocker_common_get_salt(char *ciphertext)
        p = strtokm(NULL, "$"); // data_size
        cs->data_size = atoi(p);
        p = strtokm(NULL, "$"); // data
+       for (j = 0; j < MACLEN; j++)
+               cs->mac[j] = atoi16[ARCH_INDEX(p[j * 2])] * 16
+                       + atoi16[ARCH_INDEX(p[j * 2 + 1])];
+#ifdef HAVE_OPENCL // dirty hack, will be removed later?
+       for (j = 0, i = MACLEN; i < cs->data_size; j++, i++)
+               cs->data[j] = atoi16[ARCH_INDEX(p[i * 2])] * 16
+                       + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+#else
        for (i = 0; i < cs->data_size; i++)
                cs->data[i] = atoi16[ARCH_INDEX(p[i * 2])] * 16
                        + atoi16[ARCH_INDEX(p[i * 2 + 1])];
+#endif
 
        MEM_FREE(keeptr);
```

The idea is to not change the data layout for the CPU format.

**Comment by kholia** [comment]
@e-ago The code in `opencl/bitlocker_kernel.cl` looks unreadable, and unmaintainable to me. You are probably one of the few people who can read, and debug it :-)

We already have support for AES in our OpenCL "framework". Please see how `opencl/fvde_kernel.cl` kernel is making use of the existing AES OpenCL support. AES ECB encryption/decryption can be done in just two lines.

Doing something similar would greatly simplify your kernel, and will improve its maintainability.

**Comment by kholia** [comment]
@e-ago Please add support for variable iteration count in the OpenCL code. Such a feature will make the format a bit more future work. More importantly, it will make testing, and development really easier. Currently, testing the OpenCL format takes more than five minutes (?).

Take a look at `bitlocker_kdf` in `bitlocker_fmt_plug.c` to see how it is using the variable iteration count.

https://github.com/magnumripper/JohnTheRipper/issues/3027 is also relevant to this topic.

**Comment by kholia** [comment]
I have cleaned up this PR, and created a new PR https://github.com/magnumripper/JohnTheRipper/pull/3025 which passes CI testing.

Please use my PR as the base for making improvements (i.e. variable iteration count, usage of existing AES ECB code). Thanks!

**Comment by e-ago** [comment]
Hi @kholia 

I received a bunch of notifications related to this PR! :)
In order to fully understand, I would like to recap with you:

PR https://github.com/magnumripper/JohnTheRipper/pull/3023 

Starting from the basic bitlocker2john.c file you applied my changes in without the output files. We should add a short description of every hash in order to explain that:

- hash with $0$ means User password fast attack
- hash with $1$ means User password with MAC verification
- hash with $2$ means Recovery password fast attack
- hash with $3$ means Recovery password with MAC verification

I will do it in the Wiki page, but probably we should add a description in bitlocker2john.c too.
You already merged this PR in john.

PR https://github.com/magnumripper/JohnTheRipper/pull/3025

You applied all of my changes (common and opencl_bitlocker* files).

So basically I should close this PR, review your PR https://github.com/magnumripper/JohnTheRipper/pull/3025 (from kholia:BitLocker-NG-2) and apply the iteration count improvement when you will merge https://github.com/magnumripper/JohnTheRipper/pull/3025 .
I use this "cryptic" code for AES because it's a translation from the CUDA code that I have fully optimized (for performance reasons) in my repo (register usage, local memory, etc...)

Currently, CPU code supports only User password fast attack while the OpenCL code supports both User and Recovery Password attacks with and without MAC verification.

Am I missing something?

**Comment by kholia** [comment]
Your summary is correct.

> I will do it in the Wiki page, but probably we should add a description in bitlocker2john.c too.

Sounds good. I can add a short message (e.g. `User password fast attack` and similar) before printing each hash. If you have a better idea, please let us know.

> So basically I should close this PR, review your PR #3025 (from kholia:BitLocker-NG-2) and apply the iteration count improvement (#3027) when you will merge #3025.

Yes, the next step would be to review PR #3025. It should be doable quickly as I didn't make any significant change to your OpenCL code.

> I use this "cryptic" code for AES because it's a translation from the CUDA code that I have fully optimized (for performance reasons) in my repository.

Hmm. I wonder how much difference does it really make? If you have some free time, and motivation, it would be great to give our AES ECB code a try. It can't be very bad ;)

> Currently, CPU code supports only User password fast attack while the OpenCL code supports both User and Recovery Password attacks with and without MAC verification.

Yes. To add support for other attack modes, I will need a high level description of the steps involved in the other attack modes from you. Thanks!

**Comment by kholia** [comment]
PR #3025 is merged now.

**Comment by e-ago** [comment]
Closed for https://github.com/magnumripper/JohnTheRipper/pull/3025
New updates in a different PR
